tutor/hooks/actions.py

```diff
@@ -4,58 +4,83 @@
 import sys
 import typing as t
 
+from typing_extensions import ParamSpec
+
 from .contexts import Contextualized
 
-# Similarly to CallableFilter, it should be possible to refine the definition of
-# CallableAction in the future.
-CallableAction = t.Callable[..., None]
+P = ParamSpec("P")
+# Similarly to CallableFilter, it should be possible to create a CallableAction alias in
+# the future.
+# CallableAction = t.Callable[P, None]
 
 DEFAULT_PRIORITY = 10
 
 
-class ActionCallback(Contextualized):
+class ActionCallback(Contextualized, t.Generic[P]):
     def __init__(
         self,
-        func: CallableAction,
+        func: t.Callable[P, None],
         priority: t.Optional[int] = None,
     ):
         super().__init__()
         self.func = func
         self.priority = priority or DEFAULT_PRIORITY
 
     def do(
-        self, *args: t.Any, context: t.Optional[str] = None, **kwargs: t.Any
+        self,
+        *args: P.args,
+        **kwargs: P.kwargs,
     ) -> None:
-        if self.is_in_context(context):
-            self.func(*args, **kwargs)
+        self.func(*args, **kwargs)
 
 
-class Action:
+class Action(t.Generic[P]):
     """
-    Each action is associated to a name and a list of callbacks, sorted by
-    priority.
+    Action hooks are callbacks that are triggered independently from one another.
+
+    Several actions are defined across the codebase. Each action is given a unique name.
+    To each action are associated zero or more callbacks, sorted by priority.
+
+    This is the typical action lifecycle:
+
+    1. Create an action with method :py:meth:`get` (or function :py:func:`get`).
+    2. Add callbacks with method :py:meth:`add` (or function :py:func:`add`).
+    3. Call the action callbacks with method :py:meth:`do` (or function :py:func:`do`).
+
+    The `P` type parameter of the Action class corresponds to the expected signature of
+    the action callbacks. For instance, `Action[[str, int]]` means that the action
+    callbacks are expected to take two arguments: one string and one integer.
+
+    This strong typing makes it easier to quickly check whether hooks are implemented
+    correctly by plugin developers.
```

Comment:
It's obvious to me what you mean here, but newer Tutor devs might be tripped up if we're not consistent with our terminology.
```suggestion
    This strong typing makes it easier for plugin developers to quickly check whether
    they are adding and calling action callbacks correctly.
```